File: app/CrosswordGrid.php
<?php
/***************************************************************************
 *  CrosswordGrid — Backtracking UTF-8 (LTR & RTL) — PHP ≥ 7.4
 *  
 *  • build(array $pairs, int $minWords) ⇒ self | false  
 *  • export() ⇒ array [ $cells, $across, $down ]  
 *  
 *  Chaque mot (après le premier) doit croiser au moins une lettre existante.  
 *  TIME_LIMIT (60 s) protège contre les recherches trop longues.  
 ***************************************************************************/

final class CrosswordGrid
{
    private int   $n;
    private bool  $rtl;
    private array $grid;    // [y][x] = null|string
    private array $placed;  // liste des mots placés
    private array $pairs;   // couples ['word','clue']
    private array $coords;  // positions valides par mot
    private float $start;   // timestamp du début du DFS

    private const TIME_LIMIT = 60;  // secondes max pour le DFS

    public function __construct(int $size = 15, bool $rtl = false)
    {
        $this->n   = $size;
        $this->rtl = $rtl;
        $this->grid = array_fill(0, $size, array_fill(0, $size, null));
    }

    /*======================== build() ========================*/
    public function build(array $pairs, int $minWords = 10)
    {
        $this->start  = microtime(true);

        // Trier les mots du plus long au plus court
        usort($pairs, static fn($a,$b) =>
            mb_strlen($b['word'],'UTF-8') <=> mb_strlen($a['word'],'UTF-8')
        );
        $this->pairs  = $pairs;
        $this->placed = [];

        // Pré-calculer toutes les positions possibles
        $this->coords = [];
        foreach ($pairs as $i => $p) {
            $pos = $this->allPositions($p['word']);
            if (empty($pos)) {
                // Impossible de placer ce mot
                return false;
            }
            $this->coords[$i] = $pos;
        }

        // Lancer la DFS pour placer >= $minWords mots
        return $this->dfs($minWords) ? $this : false;
    }

    /*======================== export() =======================*/
    public function export(): array
    {
        // 1) Numérotation des cases de départ
        $map = []; $num = 1;
        for ($y = 0; $y < $this->n; $y++) {
            for ($x = 0; $x < $this->n; $x++) {
                if ($this->grid[$y][$x] !== null
                    && ($this->isStart($x,$y,'across') || $this->isStart($x,$y,'down')))
                {
                    $map["{$y}-{$x}"] = $num++;
                }
            }
        }

        // 2) Construire les listes across / down
        $across = [];
        $down   = [];
        foreach ($this->placed as $p) {
            $entry = [
                'num'     => $map["{$p['y']}-{$p['x']}"],
                'answer'  => $p['word'],
                'clue'    => $p['clue'],
                'pattern' => '(' . mb_strlen($p['word'],'UTF-8') . ')',
                'dir'     => $p['dir'],
                'x'       => $p['x'],
                'y'       => $p['y'],
            ];
            if ($p['dir'] === 'across') {
                $across[] = $entry;
            } else {
                $down[]   = $entry;
            }
        }
        usort($across, static fn($a,$b)=> $a['num'] <=> $b['num']);
        usort($down,   static fn($a,$b)=> $a['num'] <=> $b['num']);

        // 3) Rogner la grille au rectangle minimal contenant tous les mots
        [$minR,$minC,$maxR,$maxC] = $this->bounds($this->placed);
        $cells = [];
        for ($y = $minR; $y <= $maxR; $y++) {
            $row = [];
            for ($x = $minC; $x <= $maxC; $x++) {
                $row[] = $this->grid[$y][$x];
            }
            $cells[] = $row;
        }

        // 4) Ajuster coordonnées (shift)
        foreach ($across as &$a) { $a['x'] -= $minC; $a['y'] -= $minR; }
        foreach ($down   as &$d) { $d['x'] -= $minC; $d['y'] -= $minR; }

        return [$cells, $across, $down];
    }

    /*=========================== dfs() ==========================*/
    private function dfs(int $goal): bool
    {
        // Timeout
        if (microtime(true) - $this->start > self::TIME_LIMIT) {
            return false;
        }
        // Objectif atteint ?
        if (count($this->placed) >= $goal) {
            return true;
        }

        // Choisir le mot avec le moins d'options de placement
        $best   = null;
        $fewest = PHP_INT_MAX;
        foreach ($this->coords as $i => $pos) {
            if (!isset($this->pairs[$i]['used']) && count($pos) < $fewest) {
                $fewest = count($pos);
                $best   = $i;
            }
        }
        if ($best === null) {
            return false;
        }

        $word = $this->pairs[$best]['word'];
        $clue = $this->pairs[$best]['clue'];

        // Essayer chaque position valide
        foreach ($this->coords[$best] as [$x,$y,$dir]) {
            if (!$this->fit($word,$x,$y,$dir)) {
                continue;
            }
            // Placer
            $this->place($word,$x,$y,$dir);
            $this->pairs[$best]['used'] = true;
            $this->placed[] = [
                'word'=>$word,'clue'=>$clue,'x'=>$x,'y'=>$y,'dir'=>$dir
            ];

            if ($this->dfs($goal)) {
                return true;
            }

            // Backtrack
            $this->remove($word,$x,$y,$dir);
            array_pop($this->placed);
            unset($this->pairs[$best]['used']);
        }

        return false;
    }

    /*--------------- allPositions() ----------------------------*/
    private function allPositions(string $w): array
    {
        $L    = mb_strlen($w,'UTF-8');
        $list = [];
        for ($y = 0; $y < $this->n; $y++) {
            for ($x = 0; $x < $this->n; $x++) {
                if ($this->fit($w,$x,$y,'across')) {
                    $list[] = [$x,$y,'across'];
                }
                if ($this->fit($w,$x,$y,'down')) {
                    $list[] = [$x,$y,'down'];
                }
            }
        }
        return $list;
    }

    /*---------------- place() -------------------------------*/
    private function place(string $w,int $x,int $y,string $d): void
    {
        $L = mb_strlen($w,'UTF-8');
        for ($i = 0; $i < $L; $i++) {
            $cx = $d==='across'
                ? ($this->rtl ? $x-$i : $x+$i)
                : $x;
            $cy = $d==='down'
                ? $y+$i
                : $y;
            $this->grid[$cy][$cx] = $this->char($w,$i,$d);
        }
    }

    /*---------------- remove() ------------------------------*/
    private function remove(string $w,int $x,int $y,string $d): void
    {
        $L = mb_strlen($w,'UTF-8');
        for ($i = 0; $i < $L; $i++) {
            $cx = $d==='across'
                ? ($this->rtl ? $x-$i : $x+$i)
                : $x;
            $cy = $d==='down'
                ? $y+$i
                : $y;

            // conserver si partagée par un autre mot
            $keep = false;
            foreach ($this->placed as $p) {
                $LL = mb_strlen($p['word'],'UTF-8');
                for ($k = 0; $k < $LL; $k++) {
                    $px = $p['dir']==='across'
                        ? ($this->rtl ? $p['x']-$k : $p['x']+$k)
                        : $p['x'];
                    $py = $p['dir']==='down'
                        ? $p['y']+$k
                        : $p['y'];
                    if ($px===$cx && $py===$cy) {
                        $keep = true;
                        break 2;
                    }
                }
            }
            if (!$keep) {
                $this->grid[$cy][$cx] = null;
            }
        }
    }

    /*---------------- char() -------------------------------*/
    private function char(string $w,int $i,string $d): string
    {
        return ($this->rtl && $d==='across')
            ? mb_substr($w,-1-$i,1,'UTF-8')
            : mb_substr($w,   $i,  1,'UTF-8');
    }

    /*---------------- fit() -------------------------------*/
    private function fit(string $w,int $x,int $y,string $d): bool
    {
        $L = mb_strlen($w,'UTF-8');
        // bordures
        if ($d==='across') {
            if (!$this->rtl && $x+$L > $this->n) return false;
            if ( $this->rtl && $x-$L+1 < 0)      return false;
        } else {
            if ($y+$L > $this->n) return false;
        }
        $over = 0;
        for ($i = 0; $i < $L; $i++) {
            $cx = $d==='across'
                ? ($this->rtl ? $x-$i : $x+$i)
                : $x;
            $cy = $d==='down'
                ? $y+$i
                : $y;
            $cell = $this->grid[$cy][$cx];
            $c    = $this->char($w,$i,$d);
            if ($cell !== null) {
                if ($cell !== $c) return false;
                $over++;
            } else {
                // pas d'adjacence parallèle
                if ($d==='across' &&
                   (($cy>0            && $this->grid[$cy-1][$cx])||
                    ($cy+1<$this->n   && $this->grid[$cy+1][$cx]))) return false;
                if ($d==='down'  &&
                   (($cx>0            && $this->grid[$cy][$cx-1])||
                    ($cx+1<$this->n   && $this->grid[$cy][$cx+1]))) return false;
            }
        }
        // doit croiser si non premier mot
        if (!empty($this->placed) && $over===0) {
            return false;
        }
        // cases avant/après
        if ($d==='across') {
            $prev = $this->rtl
                ? ($x+1<$this->n?$this->grid[$y][$x+1]:null)
                : ($x-1>=0      ?$this->grid[$y][$x-1]:null);
            $next = $this->rtl
                ? ($x-$L>=0     ?$this->grid[$y][$x-$L]:null)
                : ($x+$L<$this->n?$this->grid[$y][$x+$L]:null);
            if ($prev || $next) return false;
        } else {
            if (($y>0 && $this->grid[$y-1][$x])||
                ($y+$L<$this->n&&$this->grid[$y+$L][$x])) return false;
        }
        return true;
    }

    /*---------------- isStart() ----------------------------*/
    private function isStart(int $x,int $y,string $d): bool
    {
        if ($d==='across') {
            $left  = $this->rtl
                ? ($x==$this->n-1 || $this->grid[$y][$x+1]===null)
                : ($x==0           || $this->grid[$y][$x-1]===null);
            $right = $this->rtl
                ? ($x>0 && $this->grid[$y][$x-1]!==null)
                : ($x+1<$this->n && $this->grid[$y][$x+1]!==null);
            return $left && $right;
        }
        // down
        $up   = ($y==0 || $this->grid[$y-1][$x]===null);
        $down = ($y+1<$this->n && $this->grid[$y+1][$x]!==null);
        return $up && $down;
    }

    /*---------------- bounds() -----------------------------*/
    private function bounds(array $pl): array
    {
        $minR=$minC=$this->n;
        $maxR=$maxC=0;
        foreach ($pl as $p) {
            $L = mb_strlen($p['word'],'UTF-8');
            $minR = min($minR, $p['y']);
            $maxR = max($maxR,
                $p['dir']==='down' ? $p['y']+$L-1 : $p['y']
            );
            if ($p['dir']==='across') {
                if ($this->rtl) {
                    $minC = min($minC, $p['x']-$L+1);
                    $maxC = max($maxC, $p['x']);
                } else {
                    $minC = min($minC, $p['x']);
                    $maxC = max($maxC, $p['x']+$L-1);
                }
            } else {
                $minC = min($minC, $p['x']);
                $maxC = max($maxC, $p['x']);
            }
        }
        return [$minR,$minC,$maxR,$maxC];
    }
}

File: app/crossword_grid.py
# crossword_grid.py
import time

class CrosswordGrid:
    def __init__(self, size=15, rtl=False, time_limit=60):
        self.n = size
        self.rtl = rtl
        self.grid = [[None]*size for _ in range(size)]
        self.pairs = []
        self.placed = []
        self.coords = {}
        self.start = None
        self.time_limit = time_limit

    def build(self, pairs, min_words=10):
        # Trier du plus long au plus court
        self.pairs = sorted(pairs, key=lambda p: len(p['word']), reverse=True)
        self.placed = []
        self.start = time.time()
        self.coords = {}
        # Pré-calcul positions
        for i, p in enumerate(self.pairs):
            pos = self._all_positions(p['word'])
            if not pos:
                return False
            self.coords[i] = pos
        return self._dfs(min_words)

    def export(self):
        # Numérotation
        num = 1
        mapnum = {}
        for y in range(self.n):
            for x in range(self.n):
                if self.grid[y][x] and (self._is_start(x,y,'across') or self._is_start(x,y,'down')):
                    mapnum[f"{y}-{x}"] = num; num += 1

        across, down = [], []
        for p in self.placed:
            entry = {
                'num': mapnum[f"{p['y']}-{p['x']}"],
                'answer': p['word'],
                'clue': p['clue'],
                'pattern': f"({len(p['word'])})",
                'dir': p['dir'],
                'x': p['x'],
                'y': p['y']
            }
            (across if p['dir']=='across' else down).append(entry)

        across.sort(key=lambda e: e['num'])
        down.sort(key=lambda e: e['num'])

        # Rogner
        min_r = min(p['y'] for p in self.placed)
        max_r = max(p['y'] + (len(p['word'])-1 if p['dir']=='down' else 0) for p in self.placed)
        min_c = min(p['x'] - (len(p['word'])-1 if (self.rtl and p['dir']=='across') else 0) for p in self.placed)
        max_c = max(p['x'] + (len(p['word'])-1 if (not self.rtl and p['dir']=='across') else 0) for p in self.placed)

        cells = [
            [ self.grid[y][x] for x in range(min_c, max_c+1) ]
            for y in range(min_r, max_r+1)
        ]

        for lst in (across, down):
            for e in lst:
                e['x'] -= min_c
                e['y'] -= min_r

        return cells, across, down

    def _dfs(self, goal):
        if time.time() - self.start > self.time_limit:
            return False
        if len(self.placed) >= goal:
            return True

        # Choisir mot avec moins d’options
        choices = [i for i in self.coords if not self.pairs[i].get('used')]
        best = min(choices, key=lambda i: len(self.coords[i]), default=None)
        if best is None:
            return False

        w, clue = self.pairs[best]['word'], self.pairs[best]['clue']
        for x, y, dir in self.coords[best]:
            if not self._fit(w, x, y, dir):
                continue
            self._place(w, x, y, dir)
            self.pairs[best]['used'] = True
            self.placed.append({'word':w,'clue':clue,'x':x,'y':y,'dir':dir})

            if self._dfs(goal):
                return True

            # backtrack
            self._remove(w, x, y, dir)
            self.placed.pop()
            del self.pairs[best]['used']

        return False

    def _all_positions(self, w):
        res, L = [], len(w)
        for y in range(self.n):
            for x in range(self.n):
                for dir in ('across','down'):
                    if self._fit(w, x, y, dir):
                        res.append((x, y, dir))
        return res

    def _place(self, w, x, y, dir):
        for i, ch in enumerate(self._chars(w, dir)):
            cx = (x - i) if (self.rtl and dir=='across') else (x + i if dir=='across' else x)
            cy = (y + i) if dir=='down' else y
            self.grid[cy][cx] = ch

    def _remove(self, w, x, y, dir):
        for i, ch in enumerate(self._chars(w, dir)):
            cx = (x - i) if (self.rtl and dir=='across') else (x + i if dir=='across' else x)
            cy = (y + i) if dir=='down' else y
            keep = any(
                (cx == (p['x'] - k if (self.rtl and p['dir']=='across') else (p['x']+k if p['dir']=='across' else p['x'])))
                and (cy == (p['y']+k if p['dir']=='down' else p['y']))
                for p in self.placed for k in range(len(p['word']))
            )
            if not keep:
                self.grid[cy][cx] = None

    def _chars(self, w, dir):
        return list(w[::-1]) if (self.rtl and dir=='across') else list(w)

    def _fit(self, w, x, y, dir):
        L = len(w)
        # bordures
        if dir=='across':
            if (not self.rtl and x+L>self.n) or (self.rtl and x-L+1<0):
                return False
        else:
            if y+L>self.n:
                return False

        over = 0
        for i, ch in enumerate(self._chars(w, dir)):
            cx = (x - i) if (self.rtl and dir=='across') else (x + i if dir=='across' else x)
            cy = (y + i) if dir=='down' else y
            cell = self.grid[cy][cx]
            if cell is not None:
                if cell != ch: return False
                over += 1
            else:
                if dir=='across' and ((cy>0 and self.grid[cy-1][cx]) or (cy+1<self.n and self.grid[cy+1][cx])):
                    return False
                if dir=='down' and ((cx>0 and self.grid[cy][cx-1]) or (cx+1<self.n and self.grid[cy][cx+1])):
                    return False

        if self.placed and over==0:
            return False

        # cases avant/après
        if dir=='across':
            prev = self.grid[y][x+1] if self.rtl else (self.grid[y][x-1] if x-1>=0 else None)
            nxt  = self.grid[y][x-L] if self.rtl else (self.grid[y][x+L] if x+L<self.n else None)
            if prev or nxt: return False
        else:
            if (y>0 and self.grid[y-1][x]) or (y+L<self.n and self.grid[y+L][x]):
                return False

        return True

    def _is_start(self, x, y, dir):
        if dir=='across':
            left  = (x==self.n-1 or not self.grid[y][x+1]) if self.rtl else (x==0 or not self.grid[y][x-1])
            right = (x>0 and self.grid[y][x-1]) if self.rtl else (x+1<self.n and self.grid[y][x+1])
            return left and right
        up   = (y==0 or not self.grid[y-1][x])
        down = (y+1<self.n and self.grid[y+1][x])
        return up and down
